x/stake/keeper/slash.go

```diff
@@ -4,56 +4,212 @@ import (
 	"fmt"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	types "github.com/cosmos/cosmos-sdk/x/stake/types"
 	"github.com/tendermint/tendermint/crypto"
 )
 
-// NOTE the current slash functionality doesn't take into consideration unbonding/rebonding records
-//      or the time of breach. This will be updated in slashing v2
-// slash a validator
-func (k Keeper) Slash(ctx sdk.Context, pubkey crypto.PubKey, height int64, fraction sdk.Rat) {
+// slash an unbonding delegation and update the pool
+func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation types.UnbondingDelegation, infractionHeight int64, fraction sdk.Rat) (slashAmount sdk.Int) {
+	now := ctx.BlockHeader().Time
+
+	// If unbonding started before this height, stake didn't contribute to infraction
+	if unbondingDelegation.CreationHeight < infractionHeight {
+		return sdk.ZeroInt()
+	}
+
+	if unbondingDelegation.MinTime < now {
+		// Unbonding delegation no longer eligible for slashing, skip it
+		// TODO Settle and delete it automatically?
+		return sdk.ZeroInt()
+	}
+
+	// Calculate slash amount proportional to stake contributing to infraction
+	slashAmount = sdk.NewRatFromInt(unbondingDelegation.InitialBalance.Amount, sdk.OneInt()).Mul(fraction).EvaluateInt()
+
+	// Don't slash more tokens than held
```

Comment:
Write a note here, that this is a concern since this unbonding delegation may have already been slashed once, and we slash based on initial balance.

Reply:
Added comment, here and for the same check on redelegation.